### ISSUE TITLE
Standardize naming conventions for input schema

### ISSUE THREAD

**Issue by metasoarous** [matsengrp/olmsted#138]
This issue is a continuation of #115, the core of which has now been completed. What's left to do is finalize the schema, and I'd really like that once folks outside our lab start actually using it, we **never** break, since breakage sucks. With that in mind, it pays to spend a little bit of time at this point going through and making sure that all of the names and organization we've given here make sense for the long haul. And as suggest in matsengrp/cft#267, we should be trying (as much as possible) to make CFT match Olmsted.

Some specific concerns I have:
* [x] `asr_trees`: Many of our users may not actually have ancestral state sequence reconstructions, so this attribute name should maybe just be `trees`. (this is actually just in the _output_ of the parse script, so a little less concerned, but should still make output data and Olmsted code match input data schema where it makes sense).
* [x] `sequences`: Technically this is where we add the sequences, so this is not too bad, but it's also where we add the metadata for both tips of the tree and internal nodes, so maybe there's better language here.
* [x] `timepoint_id`: What should we be doing when one entity refers to another by id? This pattern seems reasonable, but at other points we've used the equivalent of just `timepoint`, which seems potentially confusing (cause it's not the whole timepoint). Bottom line is we should be consistent here.
* [x] `nt_seq`, `aa_seq`: Do we want to use `seq` or `sequence` for the actual sequence string (vs the record)? Again, doesn't matter too much, but want to be consistent. `dna_seq` instead?
* [x] `newick_string`: `string` seems superfluous here; Maybe `newick_tree` or `tree_newick`?
* [x] `inferred_naive_name`: This somewhat made sense from the context of CFT, but not so much here, where we really just care about what we think the naive is, not whether it was inferred or not. Suggest just `naive_name`.
* [x] `n_seqs`: We updated this in matsengrp/cft#263, and should reflect in the input schema... more generally though, we should be sure that for counts like this `n_...` is the pattern we want, and stick to it.
* [x] `sample`: Right now we nest an entire sample entity beneath every clonal family in the `clonal_families` attribute of the input `dataset` in our schema. This is not ideal from a data reusage perspective; Alternative would be to have top level `samples` attribute under dataset, and use `sample_id` or some such to make the appropriate references here. I don't _think_ we want to nest clonal families under samples, since not everyone will have things broken down the same way.
* [x] `subject`: Similar situation to sample...
* [x] `seed`: Similar to subject and sample, only moreso since this structure really only just currently contains ident and id (should have `seed_seq` or `seq` in here as appropriate?)
* [x] `reconstructions`: References back to `asr_trees` here; Is `reconstructions` a good general concept for tree+sequence+(maybe asr-seqs)? Some more general and domain appropriate?
* [x] `prune_strategy` & `prune_count`: This has to do with details about how we're downsampling clonal families so that we can build an ASR tree, which may be overly CFT specific. Still need to pass this data through, but should think about whether there's more clear/general naming we can use here.
* [x] `label`: we currently spit this out for tree nodes, but it's really just a copy of `id`, so should maybe simplify this.

**Comment by eharkins** [comment]
@metasoarous doing this now that the demand is there and I have time to spend on it. Comments:




> asr_trees: Many of our users may not actually have ancestral state sequence reconstructions, so this attribute name should maybe just be trees. (this is actually just in the output of the parse script, so a little less concerned, but should still make output data and Olmsted code match input data schema where it makes sense).

I will change this to trees

> sequences: Technically this is where we add the sequences, so this is not too bad, but it's also where we add the metadata for both tips of the tree and internal nodes, so maybe there's better language here.

`clonal_family_members`?

> timepoint_id: What should we be doing when one entity refers to another by id? This pattern seems reasonable, but at other points we've used the equivalent of just timepoint, which seems potentially confusing (cause it's not the whole timepoint). Bottom line is we should be consistent here.
 
I happy with consistent use of timepoint_id when referring to the string id and only using timepoint when referring to the "whole" timepoint record

> nt_seq, aa_seq: Do we want to use seq or sequence for the actual sequence string (vs the record)? Again, doesn't matter too much, but want to be consistent. dna_seq instead?

we mostly use `seq` I think so I will probably make things conform to that. Is dna more accurate than nt?

> newick_string: string seems superfluous here; Maybe newick_tree or tree_newick?

I can change things to use `newick_tree`

> inferred_naive_name: This somewhat made sense from the context of CFT, but not so much here, where we really just care about what we think the naive is, not whether it was inferred or not. Suggest just naive_name.

I will change to `naive_name`

>  n_seqs: We updated this in matsengrp/cft#263, and should reflect in the input schema... more generally though, we should be sure that for counts like this n_... is the pattern we want, and stick to it.

I will add unique sequences and total reads keys to the script. [Any alternative to `n_` suggestions?](https://stackoverflow.com/questions/3742650/naming-conventions-for-number-of-foos-variables)
 
>  sample: Right now we nest an entire sample entity beneath every clonal family in the clonal_families attribute of the input dataset in our schema. This is not ideal from a data reusage perspective; Alternative would be to have top level samples attribute under dataset, and use sample_id or some such to make the appropriate references here. I don't think we want to nest clonal families under samples, since not everyone will have things broken down the same way.
 
> subject: Similar situation to sample...

This makes sense, I can make the necessary changes to nest things appropriately 

>  seed: Similar to subject and sample, only moreso since this structure really only just currently contains ident and id (should have seed_seq or seq in here as appropriate?)

So the `seed` record should just be under sample, adjacent to clonal families but without anything nested in it so we just have it once? Then any seed clonal families would reference this seed record from their sample?

>  reconstructions: References back to asr_trees here; Is reconstructions a good general concept for tree+sequence+(maybe asr-seqs)? Some more general and domain appropriate?

Going based on "Many of our users may not actually have ancestral state sequence reconstructions" should we name this very generally if we still want this key to be where trees (with or without asr) are nested? Something like `clonal_family_trees` or just `trees`

>  prune_strategy & prune_count: This has to do with details about how we're downsampling clonal families so that we can build an ASR tree, which may be overly CFT specific. Still need to pass this data through, but should think about whether there's more clear/general naming we can use here.

`downsampling_strategy` and `downsampled_count`?
 > label: we currently spit this out for tree nodes, but it's really just a copy of id, so should maybe simplify this.

remove?

**Comment by eharkins** [comment]
@metasoarous #142 is done now if you have specific feedback you'd like to get from folks

**Comment by matsen** [comment]
@metasoarous I like Eli's ideas here. Thank you for your attention to detail 🕵 !

**Comment by metasoarous** [comment]
Note that we now have a live view of the schema up at http://olmstedviz.org/schema.html now thanks to @eharkins nailing #142!

Most of this looks great; I just have a few comments which I'd like feedback on from folks (@matsen @psathyrella @lauradoepker):

* How about `clonal_family_members -> tree_nodes`? This makes it clear then what the relationship is between this data and `newick_tree`.
* "Is dna more accurate than nt?": It's more specific at the very least. I don't know if you'd ever have RNA data, would you @lauradoepker @psathyrella? I've only seen DNA. I'm not at all convinced we _need_ to change this; I just want to make sure there's not some good reason we should be that anyone can think of.
* I like the SO suggestion of switching to `*_count` :+1: Shall we adopt that? 
* "So the seed record should just be under sample ...": Yes, and yes, IIUC.
* "`clonal_family_trees` or just `trees`": I thought about this as well. Glad you came to the same conclusion. I think `trees` works just fine. If we do this though, we can drop `tree` above in `tree_nodes -> nodes & newick_tree -> newick`.
* `downsampling_strategy` and `downsampled_count?`: You could use the Greek `sortition` instead of `downsampling`, if you're feeling fancy. But I like `downsampling`.
* remove `label`: Fine with this; Though I wonder if there could be use for a description, notes or some other information which might be displayed on hover. Can add later too, and drop `label` for now.

Thank you all for your feedback on this. I'm excited to see this solidify, and start getting folks using it!

**Comment by lauradoepker** [comment]
I don't follow almost any of this @metasoarous , but:
The clonal family members don't become nodes, they become leaves, right?  I'm probably missing something here.
We don't anticipate RNA, but yes, please use dna instead of nt, that's fine.
Keep "downsampling"

**Comment by metasoarous** [comment]
@lauradoepker - No worries. I can see this thread being a little hard to follow as names and interpretations are evolving upon each comment. Please let me know if there's anything I can help clarify in the final set of questions above.

> The clonal family members don't become nodes, they become leaves, right?

Remember, a node on a tree can be either a terminal node (aka leaf/tip) or an internal/node node. That is to say, `node` by itself does not imply being internal. Does that clarify things?

**Comment by eharkins** [comment]
> The clonal family members don't become nodes, they become leaves, right? I'm probably missing something here.

@lauradoepker yes, but here we are talking about clonal family members in the context of ancestral state reconstructions, and including all those ancestors as "members" of the family. Maybe that language is inaccurate, but sounds like we may not end up using it after all if we just go with `nodes`

**Comment by eharkins** [comment]
@metasoarous on board with all your and @lauradoepker's edits, let me know when I can make the changes and freeze the schema 🏔

**Comment by eharkins** [comment]
@matsen @psathyrella I am going to close unless you have any feedback. Chris and I are both satisfied with http://www.olmstedviz.org/updated_schema.html
